plutus-core/src/Language/PlutusCore/Constant/Meaning.hs

```diff
@@ -207,34 +196,59 @@ type family Merge xs ys :: [a] where
     Merge '[]       ys = ys
     Merge (x ': xs) ys = x ': Delete x (Merge xs ys)
 
--- | Collect all unique variables (a variable consists of a textual name and a unique) in an @x@.
-type family ToBinds (x :: a) :: [(Symbol, Nat)]
+-- | 'Transparent' is used to cancel 'Opaque'. We need this in cases like
+--
+--     TyAppRep (TyVarRep ('TyNameRep "f" 0)) (Transparent Integer)
+--
+-- because 'ToBinds' only handles PLC types and so if it's applied to just 'Integer', then it's
+-- stuck. The additional 'Transprent' wrapper allows us to move from handling PLC types back to
```

Comment:
> so 'Transparent' is something to "escape" from a surrounding 'Opaque'?

Yep.

> What happens if there is no surrounding Opaque? I guess it would fail

If there's no surrounding `Opaque`, then you can't put `Opaque` *inside* `Transparent`. In that case you'll get the wrong `TypeScheme` inferred and a runtime error. But you're not supposed to use `Transparent` without having a surrounding `Opaque` anyway.

I'll add a comment.